Fix: fix japanese g2p bug

litagin02 opened the pull request:
# Problem
I found that the current Japanese g2p has a bug.
For example, in the current one,
```
先生は車両を見つけると思う
```
gives
```
['_', 's', 'e', 'n', 's', 'e', 'i', 'w', 'a', 'sh', 'a', 'r', 'i', 'y', 'o', 'o', 'o', 'm', 'i', 'q', 'k', 'e', 'r', 'u', '_']
```
which is wrong, `先生` should be `センセエ` in katakana and `s e n s e e` (instead of `sensei`), and `車両` should be `シャリョオ` in katakana and `sy a ry o o` (instead of `shariyoo`), and `見つける` should be `ミツケル` in katakana and `mi ts u k e r u` (instead of `miqkeru`).
This is because in the following line:
https://github.com/fishaudio/Bert-VITS2/blob/baa61d27c0a9d2038adf378f4be9c80ded37f1e9/text/japanese.py#L35
in `kata2phoneme()`, `text`, which is already a correct pronunciation with katakana, is passed to `pyopenjtalk.g2p()`, which unfortunately convert `センセエ` to `sensei`, `シャリョオ` to `shariyoo`, and `ミツケル` to `miqkeru` (it may be a bug in pyopenjtalk, but I think pyopenjtalk.g2p() is not meant to be used with actual pronunciation kana, but used with the original text.)

Moreover, as a byproduct of this incorrect g2p, it destroys the tone: it seems that when the wrong word like `車両` or `先生` appears in text, then all the tone after that becomes 0 (left one is the original preprocessed one):

![image](https://github.com/fishaudio/Bert-VITS2/assets/139731664/65f4a74b-0119-4ff6-95fa-3c071a4c0562)


# Change
So I made a modification:
https://github.com/litagin02/Style-Bert-VITS2/blob/f04f0e14c581cf6d4a87a30ab4e13c9d49cdd5ed/text/japanese.py#L360
instead of using pyopenjtalk.g2p(), I write a new function `hiragana2p`, a little modification of jaconv's [hiragana2julius](https://github.com/ikegami-yukino/jaconv/blob/916de3a2dafe66d7bd2e74f3d7cfac45d01716f8/jaconv/jaconv.py#L526) (unfortunately I cannot find a useful library or method, so the code becomes long...), and then use it.
Then as the above image, the tone is also corrected (right one is modified one).

I confirmed that the pretrained model can be also used for training after this modification (but retraining all may improve Japanese accent), and even for models trained before this modification, the pronunciation of Japanese sound better when using this modified g2p.

I know teams are now concentrating to Chinese-specific improvement, but this may improve Japanese accent and may be important for Japanese users.

OedoSoldier commented:
Could the use of "n" rather than "N" for "ん" be more appropriate? "n" is already the established symbol for な行.

litagin02 commented:
OK I changed to use N for ん, (but may cause compatibility problem for the pretrained models, where N is not used?).
I also fixed a bug that words including `v` like `ヴァンパイア` (vampire, v a N p a i a) cannot be processed.

litagin02 commented:
Oh sorry I misread your comment and changed to use N for ん...
I think it's appropriate to use N instead of rather n, and maybe many popular Japanese TTS system uses N. but as I wrote above, this may cause compatibility problem that we cannot use pretrained models where N is not contained (I don't check actually this is a problem or not, sorry)

OedoSoldier commented:
Yes, I agree "N" is better, but we need to fine-tune the base model for the new symbol. Since we are focusing on our new project, this may take a time.
For now, we can still use "n".
